radix-engine/src/system/transaction/intent_processor.rs

```diff
@@ -387,10 +425,64 @@ impl IntentProcessorObjects {
             }
         }
 
+        if let Some(checker) = resource_constraint_checker {
+            checker.validate().map_err(|e| {
+                RuntimeError::SystemError(SystemError::IntentError(
+                    IntentError::AssertNextCallReturnsFailed(e),
+                ))
+            })?;
+        }
+
         Ok(())
     }
 }
 
+struct ResourceConstraintChecker {
+    fungible_resources: BTreeMap<ResourceAddress, Decimal>,
```

Comment:
By default, can we use IndexMap unless we have to use an invariant ordering?
`IndexMap` is more efficient than BTreeMap, and less surprising when things get arbitrarily ordered by byte-ordering of resource address.

Reply:
This was originally `IndexMap` but changed it to use `BTreeMap` since I want the ability to remove from a map without using `swap_remove`. Will use `NonIterMap` instead.

Reply:
Nothing wrong with `swap_remove`, it's just remove which explains how it affects the ordering.

But yeah, `NonIterMap` was designed for this purpose. Or if we need to iter, maybe there's some other way to do the algorithm?

Reply:
Ah good point, will switch back to `IndexMap`